concordium-node/src/consensus_ffi/ffi.rs

```diff
@@ -1665,6 +1796,233 @@ pub fn get_consensus_ptr(
+    handle: *mut DryRunHandle,
+}
+
+unsafe impl Send for DryRun {}
+
+impl Drop for DryRun {
+    fn drop(&mut self) { unsafe { dryRunEnd(self.handle) } }
+}
+
+impl DryRun {
+    /// Load the state of a particular block in the dry-run session, and use its
+    /// timestamp as the  current timestamp for the session.
+    pub fn load_block_state(
+        &mut self,
+        request: &crate::grpc2::types::BlockHashInput,
+    ) -> Result<Vec<u8>, tonic::Status> {
+        use crate::grpc2::Require;
+        let mut out_data: Vec<u8> = Vec::new();
+        let bhi = crate::grpc2::types::block_hash_input_to_ffi(request).require()?;
+        let (block_id_type, block_id) = bhi.to_ptr();
+        let res = unsafe {
+            dryRunLoadBlockState(self.handle, block_id_type, block_id.as_ptr(), &mut out_data)
+        };
+        DryRun::check_result(res, "load block state")?;
+        Ok(out_data)
+    }
+
+    /// Look up information on a particular account in the current dry-run
+    /// state.
+    pub fn get_account_info(
+        &mut self,
+        target: &crate::grpc2::types::AccountIdentifierInput,
+    ) -> Result<Vec<u8>, tonic::Status> {
+        use crate::grpc2::Require;
+        let mut out_data: Vec<u8> = Vec::new();
+        let (acc_type, acc_id) =
+            crate::grpc2::types::account_identifier_to_ffi(target).require()?;
+        let res = unsafe { dryRunGetAccountInfo(self.handle, acc_type, acc_id, &mut out_data) };
+        DryRun::check_result(res, "get account info")?;
+        Ok(out_data)
+    }
+
+    /// Look up information on a particular smart contract instance in the
+    /// current dry-run state.
+    pub fn get_instance_info(
+        &mut self,
+        target: &crate::grpc2::types::ContractAddress,
+    ) -> Result<Vec<u8>, tonic::Status> {
+        let mut out_data: Vec<u8> = Vec::new();
+        let res = unsafe {
+            dryRunGetInstanceInfo(self.handle, target.index, target.subindex, &mut out_data)
+        };
+        DryRun::check_result(res, "get instance info")?;
+        Ok(out_data)
+    }
+
+    /// Invoke an entrypoint on a smart contract instance in the current dry-run
+    /// state. No changes to the state are retained at the completion of
+    /// this operation.
+    pub fn invoke_instance(
+        &mut self,
+        request: &crate::grpc2::types::DryRunInvokeInstance,
+    ) -> Result<Vec<u8>, tonic::Status> {
+        use crate::grpc2::Require;
+        let mut out_data: Vec<u8> = Vec::new();
+
+        // Optional Address to ffi
+        let (
+            invoker_address_type,
+            invoker_account_address_ptr,
+            invoker_contract_index,
+            invoker_contract_subindex,
+        ) = if let Some(address) = &request.invoker {
+            match address.r#type.as_ref().require()? {
+                crate::grpc2::types::address::Type::Account(account) => {
+                    (1, crate::grpc2::types::account_address_to_ffi(account).require()?, 0, 0)
```

Comment:
It's not entirely clear this is ok since you're making temporary references here, then taking a pointer to those, and dropping the references.

It probably is, but usually it'd be best if the reference to `account` is retained after the pointer is needed..